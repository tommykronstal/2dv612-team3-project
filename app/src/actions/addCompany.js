import { UPDATE_COMPANY_NAME } from './types';

export function updateCompanyName (companyName) {
  return {
      type: UPDATE_COMPANY_NAME,
      companyName: companyName
  };
}

