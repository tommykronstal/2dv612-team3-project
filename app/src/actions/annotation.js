import {UPDATE_ANNOTATION} from './types'

export const updateAnnotation = (annotationText, materialId) => ({
  type: UPDATE_ANNOTATION,
  annotationText,
  materialId
})
