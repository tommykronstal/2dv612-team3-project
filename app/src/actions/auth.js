import { SET_LOGGED_IN, LOGOUT } from './types'
import { getPayloadFromJwtToken } from '../lib/jwt'

export const setUserToLoggedIn = response => ({ 
  type: SET_LOGGED_IN, 
  token: response.token, 
  payload: getPayloadFromJwtToken(response.token) 
})

export const logout = () => ({ type: LOGOUT })
