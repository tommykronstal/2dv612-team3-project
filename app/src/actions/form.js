import {CLEAR_FORM, SUBMIT_FORM, UPDATE_FIELD} from './types'

export const submitForm = (form, endpoint, action, tokenRequire = true, role = '') => ({
  type: SUBMIT_FORM,
  form,
  endpoint,
  action,
  tokenRequire,
  role
})

export const clearForm = form => ({
  type: CLEAR_FORM,
  form
})

export const updateField = (form, field, value) => ({
  type: UPDATE_FIELD,
  form,
  field,
  value
})
