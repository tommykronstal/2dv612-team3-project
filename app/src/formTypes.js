export const LOGIN = 'LOGIN'
export const ADD_COMPANY = 'ADD_COMPANY'
export const ADD_COMPANY_REP = 'ADD_COMPANY_REP'
export const REGISTER_CONSUMER = 'REGISTER_CONSUMER'
export const ADD_PRODUCT = 'ADD_PRODUCT'
export const UPDATE_PRODUCT = 'UPDATE_PRODUCT'
export const ADD_CATEGORY = 'ADD_CATEGORY'