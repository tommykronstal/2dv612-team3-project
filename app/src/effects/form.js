import {put, takeEvery, select, call} from 'redux-saga/effects'
import {SUBMIT_FORM, TOGGLE_LOADING, SET_STATUS} from '../actions/types'
import {clearForm} from '../actions/form'
import {post} from '../lib/http'
import {UPDATE_PRODUCT} from '../formTypes'

export function* watchFormActions() {
  yield takeEvery(SUBMIT_FORM, formRequest)
}

export function* formRequest({endpoint, form, action, tokenRequired}) {
  yield put({type: TOGGLE_LOADING})


  const {token, payload} = yield select(state => ({
    token: state.auth.jwt,
    payload: {
      ...state.form[form],
    },
  }))

  const tokenHeader = {...(token && tokenRequired && {Authorization: token})}

  if (form === UPDATE_PRODUCT) {
    yield sendProductFile(endpoint, token, payload, tokenHeader, action, form)
    return
  }

  console.log(payload)

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader,
    },
    body: JSON.stringify(payload)
  })

  console.log('response', response)

  yield handleResponse(response, form, action)
}

function* handleResponse(response, form, action) {
  if (response.status === 200 || response.status === 201) {
    if (action) {
      yield put(action(response))
    }
    yield put(clearForm(form))
  } else {
    if (response.message)
      yield put({type: SET_STATUS, warning: true, message: response.message})
  }

  yield put({type: TOGGLE_LOADING})
}


/**
 * Used to send data as a form.
 * Need to send it like this since original function to submit forms
 * assumes that it's JSON. (Maybe needs some tweaking in the future...)
 */
function* sendProductFile(endpoint, token, payload, tokenHeader = {}, action, form) {
  const formData = new FormData()

  console.log(payload)

  for( let key in payload) {
    formData.append(key, payload[key])
  }

  const response = yield call(post, endpoint, {
    headers: {
      ...tokenHeader
    },
    isJsonPayload: false,
    body: formData
  })

  yield handleResponse(response, form, action)
}