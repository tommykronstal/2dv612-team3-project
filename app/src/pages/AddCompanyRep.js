import React from 'react'
import {connect} from 'react-redux'

import {ADD_COMPANY_REP} from '../formTypes'
import {updateField, submitForm} from '../actions/form'
import { setStatus } from '../actions/status'
import Content from '../components/common/Content'
import Title from '../components/common/Title'
import Input from '../components/common/Input'
import Button from '../components/common/Button'

const AddCompanyRep = props => (
  <Content>
    <form onSubmit={event => { 
      event.preventDefault() 
      props.addCompanyRep()
    }}>
      <Title>Add a Company Representative</Title>
      <Input
        value={props.form.firstName}
        type='text'
        name='firstName'
        label='Representatives First Name'
        onChange={props.updateField}
      />
      <Input
        value={props.form.lastName}
        type='text'
        name='lastName'
        label='Representatives Last Name'
        onChange={props.updateField}
      />
      <Input
        value={props.form.email}
        type='email'
        name='email'
        label='Representatives E-Mail'
        onChange={props.updateField}
      />
      <Button primary loading={props.loading.isLoading}>
        Add
      </Button>
    </form>
  </Content>
)

export default connect(
  ({loading, form}) => ({
    loading, 
    form: form[ADD_COMPANY_REP] || {}
  }),
  dispatch => ({
    addCompanyRep: () => dispatch(submitForm(
      ADD_COMPANY_REP, 
      '/api/company/rep/register', 
      _ => setStatus('Company representative Created')
    )),
    updateField: ({target}) => dispatch(updateField(ADD_COMPANY_REP, target.name, target.value))   
  })
)(AddCompanyRep)
