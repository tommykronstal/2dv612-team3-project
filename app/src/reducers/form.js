import { CLEAR_FORM, UPDATE_FIELD } from '../actions/types'
import { REGISTER_CONSUMER } from '../formTypes'
import USER_TYPES from '../userTypes'

const initialState = {
  [REGISTER_CONSUMER]: {
    role: USER_TYPES.USER
  }
}

export default function status (state = initialState, action = {}) {
  switch (action.type) {
  case CLEAR_FORM:
    return {
      ...state,
      [action.form]: initialState[action.form] || {}
    }
  case UPDATE_FIELD:
    return {
      ...state,
      [action.form]: {
        ...state[action.form],
        [action.field]: action.value
      }
    }
  default:
    return state
  }
}
