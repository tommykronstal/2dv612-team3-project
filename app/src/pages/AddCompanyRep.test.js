import React from 'react'
import renderer from 'react-test-renderer'

import { AddCompanyRep, mapStateToProps, mapDispatchToProps, handleResponse } from './AddCompanyRep'
import { ADD_COMPANY_REP } from '../formTypes'
import { submitForm, updateField } from '../actions/form'

describe('AddCompanyRep', () => {

  describe('component', () => {
    it('renders correctly, initial', () => {
      const props = {
        addCompanyRep: jest.fn(),
        updateField: jest.fn(),
        form: {},
        loading: false,
        auth: { companyId: '1234' }
      }
  
      const tree = renderer.create(<AddCompanyRep {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  
    it('renders correctly, with data', () => {
      const props = {
        addCompanyRep: jest.fn(),
        updateField: jest.fn(),
        form: { 
          firstName: 'kalle',
          lastName: 'kula',
          email: 'woo@woo.com', 
          password: 'aaa' 
        },
        loading: true,
        auth: { companyId: '1234' }
      }
  
      const tree = renderer.create(<AddCompanyRep {...props} />).toJSON()
  
      expect(tree).toMatchSnapshot()
    })
  })

  describe('state', () => {
    it('should have access to company id, representative form and isLoading state', () => {
      const state = {
        auth: { companyId: '1234' },
        loading: { isLoading: false },
        form: { [ADD_COMPANY_REP]: {username: 'woooo'}}
      }
    
      expect(mapStateToProps(state)).toEqual({ 
        companyId: state.auth.companyId, 
        isLoading: state.loading.isLoading, 
        form: state.form[ADD_COMPANY_REP] 
      })
    })
  })

  describe('actions', () => {
    it('should be able to dispatch submit form actions', () => {
      const dispatch = jest.fn()
      const companyId = '1245'
    
      const props = mapDispatchToProps(dispatch)
      props.addCompanyRep(companyId)
    
      expect(dispatch).toBeCalledWith(submitForm(
        ADD_COMPANY_REP, 
        `/api/company/${companyId}`, 
        handleResponse,
      ))
    })

    it('should be able to dispatch update field actions', () => {
      const dispatch = jest.fn()
      const event = {
        target: {
          name: 'user',
          value: 'kalle'
        }
      }
    
      const props = mapDispatchToProps(dispatch)
      props.updateField(event)
    
      expect(dispatch).toBeCalledWith(updateField(ADD_COMPANY_REP, event.target.name, event.target.value))
    })
  })
})
