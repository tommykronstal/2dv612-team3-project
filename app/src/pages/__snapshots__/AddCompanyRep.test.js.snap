// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCompanyRep component renders correctly, initial 1`] = `
<div
  className="sc-htpNat iUhXvc"
>
  <div
    className="sc-bdVaJa IRNZA"
  >
    <section
      className="sc-bwzfXH cKsejz"
    >
      <form
        onSubmit={[Function]}
      >
        <h1
          className="sc-bxivhb bjAMPF"
        >
          Add a Company Representative
        </h1>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="firstName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="sc-EHOje IVybm"
            value=""
          >
            Representatives First Name
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="lastName"
            onChange={[Function]}
            type="text"
            value=""
          />
          <label
            className="sc-EHOje IVybm"
            value=""
          >
            Representatives Last Name
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="email"
            onChange={[Function]}
            type="email"
            value=""
          />
          <label
            className="sc-EHOje IVybm"
            value=""
          >
            Representatives E-Mail
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="password"
            onChange={[Function]}
            type="password"
            value=""
          />
          <label
            className="sc-EHOje IVybm"
            value=""
          >
            Password
          </label>
        </div>
        <button
          className="sc-htoDjs gyJotO"
          disabled={undefined}
          onClick={[Function]}
        >
          Add
        </button>
      </form>
    </section>
  </div>
</div>
`;

exports[`AddCompanyRep component renders correctly, with data 1`] = `
<div
  className="sc-htpNat iUhXvc"
>
  <div
    className="sc-bdVaJa IRNZA"
  >
    <section
      className="sc-bwzfXH cKsejz"
    >
      <form
        onSubmit={[Function]}
      >
        <h1
          className="sc-bxivhb bjAMPF"
        >
          Add a Company Representative
        </h1>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="firstName"
            onChange={[Function]}
            type="text"
            value="kalle"
          />
          <label
            className="sc-EHOje iRLQDf"
            value="kalle"
          >
            Representatives First Name
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="lastName"
            onChange={[Function]}
            type="text"
            value="kula"
          />
          <label
            className="sc-EHOje iRLQDf"
            value="kula"
          >
            Representatives Last Name
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="email"
            onChange={[Function]}
            type="email"
            value="woo@woo.com"
          />
          <label
            className="sc-EHOje iRLQDf"
            value="woo@woo.com"
          >
            Representatives E-Mail
          </label>
        </div>
        <div
          className="sc-bZQynM inrEaE"
        >
          <input
            className="sc-ifAKCX fDibjL"
            name="password"
            onChange={[Function]}
            type="password"
            value="aaa"
          />
          <label
            className="sc-EHOje iRLQDf"
            value="aaa"
          >
            Password
          </label>
        </div>
        <button
          className="sc-htoDjs gyJotO"
          disabled={undefined}
          onClick={[Function]}
        >
          Add
        </button>
      </form>
    </section>
  </div>
</div>
`;
